lib/auth/webauthn/register.go

```diff
@@ -0,0 +1,218 @@
+	GetWebauthnSessionData(ctx context.Context, user, sessionID string) (*wantypes.SessionData, error)
+	DeleteWebauthnSessionData(ctx context.Context, user, sessionID string) error
+}
+
+// RegistrationFlow represents the WebAuthn registration ceremony.
+//
+// Registration consists of:
+//
+// 1. Client requests a CredentialCreation (containing a challenge and various
+//    settings that may constrain allowed authenticators).
+// 2. Server runs Begin(), generates a credential creation.
+// 3. Client validates the credential creation, performs a user presence test
+//    (usually by asking the user to touch a secure token), and replies with a
+//    CredentialCreationResponse (containing the signed challenge and
+//    information about the credential and authenticator)
+// 4. Server runs Finish()
+// 5. If all server-side checks are successful, then registration is complete
+//    and the authenticator may now be used to login.
+type RegistrationFlow struct {
+	Webauthn *types.Webauthn
+	Identity RegistrationIdentity
+}
+
+// Begin is the first step of the registration ceremony.
+// The CredentialCreation created is relayed back to the client, who in turn
+// performs a user presence check and signs the challenge contained within it.
+// As a side effect Begin may assign (and record in storage) a WebAuthn ID for
+// the user.
+func (f *RegistrationFlow) Begin(ctx context.Context, user string) (*CredentialCreation, error) {
+	switch {
+	case f.Webauthn.Disabled:
+		return nil, trace.BadParameter("webauthn disabled")
+	case user == "":
+		return nil, trace.BadParameter("user required")
+	}
+
+	// Exclude known devices from the ceremony.
+	devices, err := f.Identity.GetMFADevices(ctx, user, false /* withSecrets */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var exclusions []protocol.CredentialDescriptor
+	for _, dev := range devices {
+		cred, ok := deviceToCredential(dev, true /* idOnly */)
+		if !ok {
+			continue
+		}
+		exclusions = append(exclusions, protocol.CredentialDescriptor{
+			Type:         protocol.PublicKeyCredentialType,
+			CredentialID: cred.ID,
+		})
+	}
+
+	webID, err := getOrCreateUserWebauthnID(ctx, user, f.Identity)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	u := newWebUser(user, webID, true /* credentialIDOnly */, nil /* devices */)
+
+	web, err := newWebAuthn(f.Webauthn, f.Webauthn.RPID, "" /* origin */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credentialCreation, sessionData, err := web.BeginRegistration(u, wan.WithExclusions(exclusions))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sessionDataPB, err := sessionToPB(sessionData)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := f.Identity.UpsertWebauthnSessionData(ctx, user, registrationSessionID, sessionDataPB); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return (*CredentialCreation)(credentialCreation), nil
```

Comment:
Why is this cast necessary?

Reply:
Casting from (duo-labs)wan.CredentialCreation to our (lib.auth.)webauthn.CredentialCreation. The latter is a type based on the former.

Reply:
Got it, thanks!